python/lsst/daf/butler/core/datastoreRecordData.py

```diff
@@ -105,14 +157,13 @@ def subset(self, dataset_ids: AbstractSet[DatasetId]) -> Optional[DatastoreRecor
         record_data : `DatastoreRecordData` or `None`
             `None` is returned if there are no matching refs.
         """
-        matching_refs = [ref for ref in self.refs if ref.id in dataset_ids]
-        records: Dict[str, List[StoredDatastoreItemInfo]] = {}
-        for table_name, record_data in self.records.items():
-            matching_records = [record for record in record_data if record.dataset_id in dataset_ids]
-            if matching_records:
-                records[table_name] = matching_records
-        if matching_refs or matching_records:
-            return DatastoreRecordData(refs=matching_refs, records=records)
+        matching_ids = self.dataset_ids & dataset_ids
+        matching_records: Dict[DatasetId, Dict[str, List[StoredDatastoreItemInfo]]] = {}
+        for dataset_id in matching_ids:
+            if (id_records := self.records.get(dataset_id)) is not None:
+                matching_records[dataset_id] = id_records
```

Comment:
Should we be worried about the aliasing of mutable data structures here?

I suspect not in practice, but it should probably be called out in docs.